// homey-wifidriver/index.js
'use strict';

// TODO: mdns-js
// TODO: udp-sockets
// TODO: udp-discovery
// TODO: ip scan discovery
// TODO: add support for multiple OAuth2 accounts (by user input in pairing wizard?)
// TODO: handle logging levels better (api requests, api call queue, oauth2 etc)
// TODO: add webAPIRequest error handler to OAuth2Device that will refresh tokens if unauthorized (but watch out for duplicate requests)
// TODO: add webAPIRequest middleware for OAuth2Device that checks if tokens are still valid else refresh

module.exports.WifiDevice = require('./lib/WifiDevice');
module.exports.WifiDriver = require('./lib/WifiDriver');
module.exports.WifiApp = require('./lib/WifiApp');

module.exports.WebAPIDevice = require('./lib/WebAPI/WebAPIDevice');
module.exports.WebAPIDriver = require('./lib/WebAPI/WebAPIDriver');
module.exports.WebAPIUtil = require('./lib/WebAPI/WebAPIUtil');

module.exports.OAuth2Device = require('./lib/OAuth2/OAuth2Device');
module.exports.OAuth2Driver = require('./lib/OAuth2/OAuth2Driver');
module.exports.OAuth2App = require('./lib/OAuth2/OAuth2App');
module.exports.OAuth2Util = require('./lib/OAuth2/OAuth2Util');
